File: 백준/Silver/4937. It’s All About Three/It’s All About Three.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Main {
    public static void main(String[] args) throws NumberFormatException, IOException {
	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	
	List<Integer> primes = getPrimes();
	
	int n;
	while (true) {
	    n = Integer.parseInt(br.readLine().trim());
	    if (n == -1) {
		return;
	    }
	    
	    Set<Integer> factors = new LinkedHashSet<>();
	    int idx = 0;
	    int curN = n;
	    while (curN > 1) {
		if (curN % primes.get(idx) != 0) {
		    idx++;
		} else {
		    factors.add(primes.get(idx));
		    curN /= primes.get(idx);
		}
	    }
	    
	    boolean res = true;
	    for (Integer factor : factors) {
		if (!factor.toString().endsWith("3")) {
		    res = false;
		    break;
		}
	    }
	    if (res) {
		System.out.println(n + " YES");
	    } else {
		System.out.println(n + " NO");
	    }
	}

    }

    public static List<Integer> getPrimes() {
	// 에라토스테네스의 체 생성
	Map<Integer, Boolean> eratos = new LinkedHashMap();
	for (int i = 2; i < 1_000_001; i++) {
	    eratos.put(i, true);
	}

	for (int i = 2; i < 1_000_001; i++) {
	    if (eratos.get(i)) {
		for (int j = i * 2; j < 1_000_001; j += i) {
		    eratos.replace(i, false);
		}
	    }
	}

	// 소수 리스트 생성
	List<Integer> primes = new ArrayList<>();
	for (Map.Entry<Integer, Boolean> entry : eratos.entrySet()) {
	    Integer key = entry.getKey();
	    primes.add(key);
	}

	return primes;
    }
}

File: 백준/Silver/4937. It’s All About Three/It’s All About Three.py
import sys
input = sys.stdin.readline

eratos = {i: True for i in range(2, 10**6+1)}
for i in range(2, 10**6+1):
    if eratos[i]:
        for j in range(i*2, 10**6+1, i):
            eratos[j] = False

primes = []
for k, v in eratos.items():
    if v:
        primes.append(k)
# print("--- eratos completed ---")

def solve(n):
    if n == 1:
        return False

    factors = set()
    idx = 0
    while n > 1:
        if n % primes[idx] != 0:
            idx += 1
        else:
            factors.add(str(primes[idx]))
            n //= primes[idx]
    
    for factor in factors:
        if factor[-1] != "3":
            return False
        
    return True


while True:
    n = int(input())
    if n == -1:
        break
    
    if solve(n):
        print(f"{n} YES")
    else:
        print(f"{n} NO")
